### PR TITLE
Finish Unix support for X509 extensions

### PR DESCRIPTION
**Pull request by bartonjs**
With this change all of the tests in ExtensionsTests.cs are enabled (and passing) on Unix builds.

New DER encoding helpers were added
* Boolean (dead simple, only two answers :smile:)
* OID (slightly complicated, in a commit by itself)
* Octet String (dead simple)
* (non-named-list) Bit String (remarkably simple)

The `byte[][]` Segmented helpers are reducing the amount of memory copying, but also kind of a pain.  I considered trying to make rich types for them in the middle of this change, but if we can avoid too much more I think still that's better suited to public API (post-1.0.0-rtm).

This is a partial fix for #1993.  I count 8 total NotImplementedExceptions remaining in the Unix X509 code; with perhaps 6 of them moving to PlatformNotSupportedException.

cc: @stephentoub @morganbr @AtsushiKan

### CODE REVIEW

**Comment by stephentoub** [comment]
test code formatter check

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 1]
```diff
@@ -274,13 +351,129 @@ internal static byte[][] SegmentedEncodeNamedBitList(byte[] bigEndianBytes, int
...
+            };
+        }
+
+        /// <summary>
+        /// Encode the segments { tag, length, value } of an object identifier (Oid).
+        /// </summary>
+        /// <returns>The encoded segments { tag, length, value }</returns>
+        internal static byte[][] SegmentedEncodeOid(Oid oid)
+        {
+            if (oid == null)
```
Comment (AtsushiKan) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Why we do throw here for a bad argument when the pattern for other DerEcode methods is to assert?

**Comment by stephentoub** [comment]
Other than Ati's comments and the compilation errors, LGTM.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 1 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: The null check here I can probably change to an assert safely; the throws are for things like

```C#
OidCollection ekus = new OidCollection
{
    new Oid("3"), // Doesn't throw
    new Oid("Unresolved Friendly Name"), // Doesn't throw
};
var ekuExt = new X509EnhancedKeyUsageExtension(ekus);
byte[] data = ekuExt.RawData; // throws CryptographicException on Encode
```

It was the only thing that really had the notion of invalid data permeating it, so it's the only one that had to bother to throw.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 2]
```diff
@@ -80,6 +83,28 @@ private static byte[] EncodeLength(int length)
...
+            // DER says TRUE is 0xFF.
+            byte[] data =
+            {
+                (byte)(value ? 0xFF : 0x00),
+            };
+
+            return new byte[][]
+            {
+                new byte[] { (byte)DerSequenceReader.DerTag.Boolean }, 
+                new byte[] { 0x01 },
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Having to build structures like this suggests a stronger type really could be worthwhile, especially so that we can guarantee tags and lengths are right everywhere

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 1 above)
Comment (AtsushiKan) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: That’s fine – that specific null check was the one that I was commenting on. Stuff that the caller can’t reasonably check is fine to throw on.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 3]
```diff
@@ -149,6 +174,58 @@ internal static byte[][] SegmentedEncodeUnsignedInteger(byte[] bigEndianBytes, i
...
+            // We need to make a mask of the bits to keep around for the last
+            // byte, ensuring we clear out any bits that were set, but reported
+            // as padding by unusedBits
+            // 
+            // For example:
+            // unusedBits 0 => mask 0b10000000
+            // unusedBits 1 => mask 0b11000000
+            // unusedBits 7 => mask 0b11111111
+            // unusedBits 8 => mask 0b10000000
+            byte lastByteSemanticMask = unchecked((byte)(-1 << unusedBits));
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: This doesn't seem to match the examples above (-1 << 0 should be 0xFF)

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 2 above)
Comment (AtsushiKan) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: >Having to build structures like this suggests a stronger type really could be worthwhile, especially so that we can guarantee tags and lengths are right everywhere

Seconded. (though it’d be ok to address in a separate PR as this one’s quite substantial already.)

Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Or at a minimum having a common method that takes a tag and data

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 4]
```diff
@@ -274,13 +351,129 @@ internal static byte[][] SegmentedEncodeNamedBitList(byte[] bigEndianBytes, int
...
+                throw new ArgumentNullException("oid");
+
+            // All exceptions past this point should just be "CryptographicException", because that's
+            // how they'd come back from Desktop/Windows, since it was a non-success result of calling
+            // CryptEncodeObject.
+            string oidValue = oid.Value;
+
+            if (string.IsNullOrEmpty(oidValue))
+                throw new CryptographicException();
+            if (oidValue.Length < 3 /* "1.1" is the shortest vlaue */)
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Nit -- typo

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 2 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: I'd like to take zeroth credit:

> The byte[][] Segmented helpers are reducing the amount of memory copying, but also kind of a pain. I considered trying to make rich types for them in the middle of this change, but if we can avoid too much more I think still that's better suited to public API (post-1.0.0-rtm).

Since right now it's a source-include I'm trying to keep it light; but maybe some nested types are called for.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 5]
```diff
@@ -274,13 +351,129 @@ internal static byte[][] SegmentedEncodeNamedBitList(byte[] bigEndianBytes, int
...
+        /// <summary>
+        /// Encode the segments { tag, length, value } of an object identifier (Oid).
+        /// </summary>
+        /// <returns>The encoded segments { tag, length, value }</returns>
+        internal static byte[][] SegmentedEncodeOid(Oid oid)
+        {
+            if (oid == null)
+                throw new ArgumentNullException("oid");
+
+            // All exceptions past this point should just be "CryptographicException", because that's
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Should we have further info in the CryptographicExceptions? Windows would at least have an HRESULT and may have a message that goes with them

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 3 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Oops, copied it from the NamedBitList where the input value had the opposite semantics.  Will fix.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 6]
```diff
@@ -274,13 +351,129 @@ internal static byte[][] SegmentedEncodeNamedBitList(byte[] bigEndianBytes, int
...
+                    break;
+                default:
+                    throw new CryptographicException();
+            }
+
+            int startPos = 2;
+            int nextDot = oidValue.IndexOf('.', startPos);
+
+            BigInteger rid = ParseOidRid(oidValue, startPos, nextDot);
+            rid += 40 * firstRid;
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: What's 40?

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 7]
```diff
@@ -274,13 +351,129 @@ internal static byte[][] SegmentedEncodeNamedBitList(byte[] bigEndianBytes, int
...
+                    firstRid = 0;
+                    break;
+                case '1':
+                    firstRid = 1;
+                    break;
+                case '2':
+                    firstRid = 2;
+                    break;
+                default:
+                    throw new CryptographicException();
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Do we actually care about enforcing this?

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 5 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Window's message (for an OID starting with "3.") is `ASN1 bad arguments to function call` (or for the empty-string OID `The parameter is incorrect.`).

I guess I can float Argument_InvalidOidValue: `The OID value was invalid.` up into the Encodings tests so that they can continue to build the tests for this internal class.   Doesn't quite match what Windows does, but is arguably better.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 6 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Added the spec explanation/comment

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 8]
```diff
@@ -274,13 +351,129 @@ internal static byte[][] SegmentedEncodeNamedBitList(byte[] bigEndianBytes, int
...
+            };
+        }
+
         /// <summary>
         /// Make a constructed SEQUENCE of the byte-triplets of the contents.
         /// Each byte[][] should be a byte[][3] of {tag (1 byte), length (1-5 bytes), payload (variable)}.
         /// </summary>
         internal static byte[] ConstructSequence(params byte[][][] items)
         {
+            return ConstructSequence((IEnumerable<byte[][]>)items);
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Why cast here?

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 9]
```diff
@@ -324,5 +517,99 @@ internal static byte[] ConstructSequence(params byte[][][] items)
...
         }
+
+        private static BigInteger ParseOidRid(string oidValue, int startIndex, int endIndex)
+        {
+            if (startIndex > oidValue.Length)
+                throw new CryptographicException();
+
+            if (endIndex == -1)
+            {
+                endIndex = oidValue.Length;
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: This looks off by oneish. I see that it works on the loop below, but you might at least need a rename and/or comments for the parameter

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 10]
```diff
@@ -324,5 +517,99 @@ internal static byte[] ConstructSequence(params byte[][][] items)
...
+            for (position++; position < endIndex; position++)
+            {
+                value *= 10;
+                value += AtoI(oidValue[position]);
+            }
+
+            return value;
+        }
+
+        private static int AtoI(char c)
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Can you replace this with Convert.ToInt32 (or ToByte)?

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs [hunk 11]
```diff
@@ -324,5 +517,99 @@ internal static byte[] ConstructSequence(params byte[][][] items)
...
+
+            // The following code is equivalent to
+            // BigInteger.TryParse(temp, NumberStyles.None, CultureInfo.InvariantCulture, out value)
+            // but doesn't require creating the intermediate substring (TryParse doesn't take start+end/length values)
+
+            int position = startIndex;
+            int seed = AtoI(oidValue[position]);
+            BigInteger value = new BigInteger(seed);
+
+            for (position++; position < endIndex; position++)
```
Comment (morganbr) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Doing a postincrement in the declaration is pretty confusing. Could you just increment it above?

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 10 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: > Can you replace this with Convert.ToInt32 (or ToByte)?

That would return the ASCII code, not the 0-9 number.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 8 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: `internal static byte[] ConstructSequence(params byte[][][] items)` is a better match than `internal static byte[] ConstructSequence(IEnumerable<byte[][]> items)` without the cast, preferring a StackOverflowException to something functional :)

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 4 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Fixed :)

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 9 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: I updated this a bit, now it just takes `ref int startIndex` and hides its bounds checking.

File: src/Common/src/System/Security/Cryptography/DerEncoder.cs (hunk 11 above)
Comment (bartonjs) [inline src/Common/src/System/Security/Cryptography/DerEncoder.cs]: Yeah, cleaned this up a bit.

**Comment by bartonjs** [comment]
Merged the fixes into their respective smaller commits; and added a new commit to add some negative tests.

I think the Windows implementation may have regressed on the handling of the empty-string-oid-value (throws on Desktop, succeeds on CoreFx), looking into what may have happened.

**Comment by bartonjs** [comment]
@mmitche Sorry if you're inundated with notifications, but Jenkins seems to be unresponsive, and hasn't picked this job up after many hours.

**Comment by AtsushiKan** [comment]
LGTM.

**Comment by mmitche** [comment]
I'm on it.

**Comment by mmitche** [comment]
@dotnet-bot test this please
